feat 18560 refactor assessmenttable export

breierDemos opened the pull request:
### Ticket
[18560](https://www.dev.diplanung.de/DefaultCollection/BOP-HH/_workitems/edit/18560)
[18561](https://www.dev.diplanung.de/DefaultCollection/BOP-HH/_workitems/edit/18561)

This PR tries to extract some logic from the invoke function of the pdf exporter (assessment table export) into own functions.

### How to review/test
test the assessment table pdf export in projects with assessment table. code review (commit wise)

### PR Checklist
<!-- Reminders for handling PRs -->

Delete the checkbox if it doesn't apply/isn't necessary.

- [ ] Tests updated/created
- [ ] Update documentation
- [x] Link all relevant tickets
- [ ] Move the tickets on the board accordingly
- [ ] Data-Cy attributes added/updated ([conventions](https://dplan-documentation.demos-europe.eu/development/guidelines-conventions/coding-styleguides/twig_html.html#guideline-for-naming-cypress-hooks))

breierDemos committed:
refactor (refs AB#18560): Add ValueObject ExportParameters. Instead of creating an array of parameters, a ValueObject will be used. For now, it has a toArray() method to ensure the further logic is functional.

breierDemos committed:
refactor (refs AB#18560): initialize properties

breierDemos committed:
refactor (refs AB#18560): Remove unused use statement.

breierDemos committed:
refactor (refs AB#18560): Extract logic to determine list line width in own function.

breierDemos committed:
refactor (refs AB#18560): Reduce cognitively complexity of new function. Extract conditions in new functions.

breierDemos committed:
refactor (refs AB#18560): Create enum for magic numbers of list line width.

breierDemos committed:
refactor (refs AB#18560): Extract filtering of statements for condensed export to won function.

breierDemos committed:
refactor (refs AB#18560): Extract preparation of statements for condensed template in own function.

breierDemos committed:
refactor (refs AB#18560): Remove $key parameter callback, because not used. Remove else, not needed here.

breierDemos committed:
refactor (refs AB#18560): $templateVars was never updated in foreach (copy by value)

breierDemos committed:
refactor (refs AB#18560): Save the idArrays in a temporary array to fix php inspection.

breierDemos committed:
refactor (refs AB#18560): Add a ValueObject for holding fragmentIds and statementIds. Extract logic to separate statementIds and fragmentIds in own function.

breierDemos committed:
refactor (refs AB#18560): Remove old todo.

breierDemos committed:
refactor (refs AB#18560): Extract the creation of the pdf to won function.

demosci committed:
style: Apply php-cs-fixer

breierDemos committed:
feature (refs AB#18560): Remove ValueObject ExportParameters for now. Refactoring the export parameters is too complex for now.

breierDemos committed:
feature (refs AB#18560): correct use statement

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

demosci committed:
style: Apply php-cs-fixer

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
Merge branch 'main' into f-18560-refactor-assessmenttable-export

sierrazamudiodemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/AssessmentTableExporter/AssessmentTablePdfExporter.php:
```diff
@@ -626,4 +539,148 @@ private function isHorizontalNotSplitView(string $template, string $templateName
+     */
+    private function filterStatementsForCondensedExport(array $statements, string $template, string $exportType): array
+    {
+        if ('condensed' === $template && 'statementsOnly' !== $exportType) {
```
what about making condensed and statementesOnly constatns on their respective entites that are relevant and user them here?

breierDemos replied:
I extracted them into string-backed enums under .../AssessmentTableExporter/Enum/, matching the ListLineWidth/TextLineWidth precedent in this PR:
- ExportTemplate — the template parameter values (condensed, landscape, landscapeWithFrags, portrait, portraitWithFrags)
- ExportType — the exportType parameter values (statementsAndFragments, statementsOnly, fragmentsOnly)▎
- ExportTemplateName — the selected Twig template basename (export, export_condensed, export_condensed_anonymous, export_anonymous, export_original, export_fragments_anonymous)

commit: [c903e38](https://github.com/demos-europe/demosplan-core/pull/3567/commits/c903e3852103714d85b05c08dd70a4b715f2e3f0)

sierrazamudiodemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/AssessmentTableExporter/AssessmentTablePdfExporter.php:
```diff
@@ -132,26 +134,12 @@ public function __invoke(array $parameters): array
-            // out later along the way.
+            $idCollection = $this->extractStatementAndFragmentIds($parameters);
+            $fragmentIds = $idCollection->getFragmentIds();
+            if (null !== $idCollection->getStatementIds()) {
+                $parameters['filters']['id'] = $idCollection->getStatementIds();
+            }
```
Could you elaborate in what part of the old code this is happening? I still dont get it

breierDemos replied:
On main it was:

 ```
$fragmentIds = [];
if (array_key_exists('items', $parameters) && 0 < (is_countable($parameters['items']) ? count($parameters['items']) : 0)) {
      $parameters['filters']['id'] = [];
      foreach ($parameters['items'] as $statementOrFragmentId) {
          $idArray = $this->assessmentHandler
              ->getStatementIdFromStatementIdOrStatementFragmentId(
                  $statementOrFragmentId
              );
          $parameters['filters']['id'][] = $idArray['statementId'];
          if (null !== $idArray['fragmentId']) {
              $fragmentIds[] = $idArray['fragmentId'];
          }
          unset($idArray);
      }
  }
```

 items is a mixed list of statement and fragment IDs. Each entry is resolved to its parent statement ID for filters['id'] (the Elasticsearch query only indexes statements), while fragment IDs are collected separately so selected fragments can be filtered later. I moved this into extractStatementAndFragmentIds(), which returns both lists in an IdCollection instead of two loose variables.

sierrazamudiodemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/AssessmentTableExporter/AssessmentTablePdfExporter.php:
```diff
@@ -165,92 +153,35 @@ public function __invoke(array $parameters): array
-                foreach ($templateVars['table']['entries']['statements'] as $singleStatementData) {
+                foreach ($templateVars['table']['entries']['statements'] as $key => $singleStatementData) {
                     $singleStatementData['meta']['houseNumber'] = '';
+                    $templateVars['table']['entries']['statements'][$key] = $singleStatementData;
```
in what part of the code this is needed?

breierDemos replied:
meta.houseNumber is rendered into the address line of the assessment-table export templates — export_assessment_table_entry.tex.twig (line 23) and export_assessment_table_original_entry.tex.twig (line 22), e.g. {{ statement.meta.orgaStreet }} {{ statement.meta.houseNumber }} .... The feature_statement_meta_house_number_export permission controls whether the house number is allowed in the export, so when the user doesn't have it we blank meta['houseNumber'] here before the template runs — otherwise the real house number would be printed in the PDF.

You could check the pdf export with 'feature_statement_meta_house_number_export' disabled or enabled.

sierrazamudiodemos commented:
Good job on isolating the complex logic on methods.
I have 2 remarks:
- the question already reaised
- is is possible to transform some strings into constants?
Thanks for the refactor

breierDemos committed:
Merge remote-tracking branch 'origin/main' into f-18560-refactor-assessmenttable-export

# Conflicts:
#	demosplan/DemosPlanCoreBundle/Logic/Statement/AssessmentTableExporter/AssessmentTablePdfExporter.php

breierDemos committed:
refactor (refs AB#18560): Replace export template/type magic strings with enums

Introduce ExportTemplate, ExportType and ExportTemplateName enums and use
them in AssessmentTablePdfExporter and AssessmentTableFileExporterAbstract
instead of repeated string literals.

breierDemos committed:
refactor (refs AB#18560): Fix IdCollection getStatementIds nullable annotation

The statementIds property is ?array defaulting to null, but the docblock
declared it as array, causing a PHPStan notIdentical.alwaysTrue false
positive on the null guard in AssessmentTablePdfExporter.

breierDemos committed:
refactor (refs AB#18560): Use null|array syntax for getStatementIds annotation

demosci committed:
style: Apply php-cs-fixer

breierDemos committed:
refactor (refs AB#18560): Restore call to prepareStatementsForCondensedTemplate

A merge of main reintroduced the inline condensed-statement block in invoke()
and left prepareStatementsForCondensedTemplate() unused. Replace the duplicated
inline block with the call to the extracted method again.

sierrazamudiodemos commented on demosplan/DemosPlanCoreBundle/ValueObject/AssessmentTable/IdCollection.php:
```diff
@@ -0,0 +1,27 @@
+ */
+class IdCollection extends ValueObject
+{
+    protected array $fragmentIds;
```
what about initiaizing with empty array?

breierDemos replied:
Adjusted in b55c027

breierDemos committed:
Merge remote-tracking branch 'origin/main' into f-18560-refactor-assessmenttable-export

breierDemos committed:
refactor (refs AB#18560): initialize fragmentIds with empty array default

breierDemos committed:
Merge remote-tracking branch 'origin/main' into f-18560-refactor-assessmenttable-export

sierrazamudiodemos reviewed (approved):
thx a lot for all of this!
